### DISCUSSION TITLE
For testing it would be nice to have set_data method in RequestExt

### DISCUSSION THREAD

**Discussion by ry** [routerify/routerify#123] [Ideas]
How do people do write unit tests against handler that use Routerify's request data?

**Comment by rousan** [comment]
@ry 
The `RequestExt.data()` method facilitates data and app state sharing across route handlers, middleware, and the error handler in Routerify. For further information, please refer to the documentation [here](https://docs.rs/routerify/latest/routerify/#data-and-state-sharing).

Currently, there is no public API available to directly test the handler with Routerify's request data. To address this, we can propose two possible approaches:

### Approach 1: Expose `SharedDataMap` and `DataMap` internal types as follows:

```rust
#[tokio::test]
async fn test_router_handler() {
    struct Data(u64);

    // A handler for "/" page.
    async fn home_handler(req: Request<Body>) -> Result<Response<Body>, Infallible> {
        // Access the app state.
        let data = req.data::<Data>().unwrap();

        Ok(Response::new(Body::from(data.0.to_string())))
    }

    let mut req = Request::builder()
        .method("GET")
        .uri("http://localhost/example")
        .body(Body::empty())
        .unwrap();

    // Prepare data.
    let mut data_map = DataMap::new();
    data_map.insert(Data(100));
    let shared_data_maps = vec![SharedDataMap::new(Arc::new(data_map))];

    // Inject the data to Request's extensions..
    let ext = req.extensions_mut();
    ext.insert(shared_data_maps);

    let resp = home_handler(req).await.unwrap();

    let resp = into_text(resp.into_body()).await;
    assert_eq!(resp, "100");
}
```

### Approach 2: Introduce a separate method called `set_data_test()` in `RequestExt` for modifying shared data in the test environment:

```rust
#[tokio::test]
async fn test_router_handler() {
    struct Data(u64);

    // A handler for "/" page.
    async fn home_handler(req: Request<Body>) -> Result<Response<Body>, Infallible> {
        // Access the app state.
        let data = req.data::<Data>().unwrap();

        Ok(Response::new(Body::from(data.0.to_string())))
    }

    let mut req = Request::builder()
        .method("GET")
        .uri("http://localhost/example")
        .body(Body::empty())
        .unwrap();

    // Update the shared data only for this request.
    req.set_data_test(Data(100));

    let resp = home_handler(req).await.unwrap();

    let resp = into_text(resp.into_body()).await;
    assert_eq!(resp, "100");
}
```

IMO, We should go for Approach 2 and expose the `set_data_test` method only for test environment.

**Comment by avosa** [comment]
@rousan I concur.

**Comment by ta3pks** [comment]
You can always set your data container as something like `Arc<RwLock<T>>` then you can just update inside of it as *r.data().unwrap().write() = <SomeOtherT>
this would be a solution to what you are asking for.
Also you can use Any type and downcast it safely to any value this is another solution to your particular problem.

In either case this is a niche problem and I dont think this should be supported in routerify itself since it can cause more problems then it solves. As I demonstrated above routerify doesnt really prevent you from doing this by yourself but this is an app specific issue rather than being a general issue in the domain what routerify tries to solve

**Comment by rousan** [comment]
@ta3pks,
this is a good solution, but you need to write a router builder and then setup a mock server and send a mock request to the server to unit-test a handler.
@ry is trying to unit-test the handler directly by passing the `req` as an argument.

**Comment by ta3pks** [comment]
Oh i see I got it wrong then. But in this case I dont think he should be involving routerify in his unit tests since he is not testing routerify but his own application logic. I still believe this is not a case routerify should try to solve this seems a design issue with the app that utilizes routerify
